diff --git a/support.js b/support.js
index dcec20929928e06e5b50d7482e434eeda57a314b..a4a59cb7f73b1a31c4531201808d7ef2729683a4 100644
--- a/support.js
+++ b/support.js
@@ -26,7 +26,7 @@ const sendCoverage = (coverage, pathname = '/') => {
     sendBatchCoverage(totalCoverage, envBatchSize)
   } else {
     cy.task('combineCoverage', JSON.stringify(totalCoverage), {
-      log: false
+      log: true
     })
   }
 }
@@ -47,7 +47,7 @@ const sendBatchCoverage = (totalCoverage, batchSize) => {
     })
 
     cy.task('combineCoverage', JSON.stringify(batchCoverage), {
-      log: false
+      log: true
     })
   }
 }
@@ -193,10 +193,11 @@ const registerHooks = () => {
       function captureCoverage(url, suffix = '') {
         cy.request({
           url,
-          log: false,
-          failOnStatusCode: false
+          log: true,
+          failOnStatusCode: true
         })
           .then((r) => {
+            cy.log(JSON.stringify(r.body))
             return Cypress._.get(r, 'body.coverage', null)
           })
           .then((coverage) => {
