// @nx-dotnet/core/src/models/build-executor-configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetBuildExecutorConfiguration = void 0;
const devkit_1 = require("@nx/devkit");
const semver_1 = require("semver");
/**
 * Returns a TargetConfiguration for the nx-dotnet/core:build executor
 */
function GetBuildExecutorConfiguration(projectRoot) {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const outputs = (0, semver_1.lt)(devkit_1.NX_VERSION, '15.0.0-beta.0')
        ? [`dist/${projectRoot}`, `dist/intermediates/${projectRoot}`]
        : [
            `{workspaceRoot}/dist/${projectRoot}`,
            `{workspaceRoot}/dist/intermediates/${projectRoot}`,
        ];
    return {
        executor: '@nx-dotnet/core:build',
        outputs,
        options: {
            configuration: 'Debug',
            noDependencies: true,
        },
        configurations: {
            production: {
                configuration: 'Release',
            },
        },
    };
}
exports.GetBuildExecutorConfiguration = GetBuildExecutorConfiguration;
