// nx/src/command-line/release/command-object.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.yargsReleaseCommand = void 0;
const yargs_1 = require("yargs");
const file_utils_1 = require("../../project-graph/file-utils");
const shared_options_1 = require("../yargs-utils/shared-options");
exports.yargsReleaseCommand = {
    command: 'release',
    describe: '**ALPHA**: Orchestrate versioning and publishing of applications and libraries',
    builder: (yargs) => yargs
        .command(versionCommand)
        .command(changelogCommand)
        .command(publishCommand)
        .demandCommand()
        .option('groups', {
        description: 'One or more release groups to target with the current command.',
        type: 'string',
        coerce: shared_options_1.parseCSV,
        alias: ['group', 'g'],
    })
        .option('projects', {
        type: 'string',
        alias: 'p',
        coerce: shared_options_1.parseCSV,
        describe: 'Projects to run. (comma/space delimited project names and/or patterns)',
    })
        .option('dryRun', {
        describe: 'Preview the changes without updating files/creating releases',
        alias: 'd',
        type: 'boolean',
        default: false,
    })
        .option('verbose', {
        type: 'boolean',
        describe: 'Prints additional information about the commands (e.g., stack traces)',
    })
        .check((argv) => {
        if (argv.groups && argv.projects) {
            throw new Error('The --projects and --groups options are mutually exclusive, please use one or the other.');
        }
        const nxJson = (0, file_utils_1.readNxJson)();
        if (argv.groups?.length) {
            for (const group of argv.groups) {
                if (!nxJson.release?.groups?.[group]) {
                    throw new Error(`The specified release group "${group}" was not found in nx.json`);
                }
            }
        }
        return true;
    }),
    handler: async () => {
        (0, yargs_1.showHelp)();
        process.exit(1);
    },
};
const versionCommand = {
    command: 'version [specifier]',
    aliases: ['v'],
    describe: 'Create a version and release for one or more applications and libraries',
    builder: (yargs) => yargs
        .positional('specifier', {
        type: 'string',
        describe: 'Exact version or semver keyword to apply to the selected release group.',
    })
        .option('preid', {
        type: 'string',
        describe: 'The optional prerelease identifier to apply to the version, in the case that specifier has been set to prerelease.',
        default: '',
    }),
    handler: (args) => Promise.resolve().then(() => require('./version')).then((m) => m.versionHandler(args)),
};
const changelogCommand = {
    command: 'changelog [version]',
    aliases: ['c'],
    describe: 'Generate a changelog for one or more projects, and optionally push to Github',
    builder: (yargs) => yargs
        // Disable default meaning of yargs version for this command
        .version(false)
        .positional('version', {
        type: 'string',
        description: 'The version to create a Github release and changelog for',
    })
        .option('from', {
        type: 'string',
        description: 'The git reference to use as the start of the changelog. If not set it will attempt to resolve the latest tag and use that',
    })
        .option('to', {
        type: 'string',
        description: 'The git reference to use as the end of the changelog',
        default: 'HEAD',
    })
        .option('interactive', {
        alias: 'i',
        type: 'string',
        description: 'Interactively modify changelog markdown contents in your code editor before applying the changes. You can set it to be interactive for all changelogs, or only the workspace level, or only the project level',
        choices: ['all', 'workspace', 'projects'],
    })
        .option('gitRemote', {
        type: 'string',
        description: 'Alternate git remote in the form {user}/{repo} on which to create the Github release (useful for testing)',
        default: 'origin',
    })
        .option('tagVersionPrefix', {
        type: 'string',
        description: 'Prefix to apply to the version when creating the Github release tag',
        default: 'v',
    })
        .check((argv) => {
        if (!argv.version) {
            throw new Error('A target version must be specified');
        }
        if (argv.file === false && argv.createRelease !== 'github') {
            throw new Error('The --file option can only be set to false when --create-release is set to github.');
        }
        return true;
    }),
    handler: (args) => Promise.resolve().then(() => require('./changelog')).then((m) => m.changelogHandler(args)),
};
const publishCommand = {
    command: 'publish',
    aliases: ['p'],
    describe: 'Publish a versioned project to a registry',
    builder: (yargs) => (0, shared_options_1.withRunManyOptions)((0, shared_options_1.withOutputStyleOption)(yargs))
        .option('registry', {
        type: 'string',
        description: 'The registry to publish to',
    })
        .option('tag', {
        type: 'string',
        description: 'The distribution tag to apply to the published package',
    }),
    handler: (args) => Promise.resolve().then(() => require('./publish')).then((m) => m.publishHandler(coerceParallelOption((0, shared_options_1.withOverrides)(args, 2)))),
};
function coerceParallelOption(args) {
    if (args['parallel'] === 'false' || args['parallel'] === false) {
        return {
            ...args,
            parallel: 1,
        };
    }
    else if (args['parallel'] === 'true' ||
        args['parallel'] === true ||
        args['parallel'] === '') {
        return {
            ...args,
            parallel: Number(args['maxParallel'] || args['max-parallel'] || 3),
        };
    }
    else if (args['parallel'] !== undefined) {
        return {
            ...args,
            parallel: Number(args['parallel']),
        };
    }
    return args;
}
